moved/src/state_actor/queries.rs

```diff
@@ -44,67 +62,83 @@ pub trait StateQueries {
         db: &(impl TreeReader<StateKey> + Sync),
         account: AccountAddress,
         height: BlockHeight,
-    ) -> Nonce;
+    ) -> Option<Nonce>;
 }
 
 #[derive(Debug)]
 pub struct StateMemory {
     mem: InMemoryVersionedState,
-    height_to_version: HashMap<u64, u64>,
-}
-
-impl Default for StateMemory {
-    fn default() -> Self {
-        Self::new()
-    }
+    tags: Vec<Version>,
+    version: Version,
 }
 
 impl StateMemory {
-    pub fn from_genesis(changes: ChangeSet) -> Self {
+    /// Creates state memory with `genesis_changes` on `version` 0 tagged as block `height` 0.
+    pub fn from_genesis(genesis_changes: ChangeSet) -> Self {
         Self {
-            mem: InMemoryVersionedState::from_iter(Self::convert(changes, 0)),
-            height_to_version: HashMap::from([(0, 0)]),
+            mem: InMemoryVersionedState::from_iter(genesis_changes.into_versioned_kv(0)),
+            tags: vec![0],
+            version: 0,
         }
     }
 
-    pub fn new() -> Self {
-        Self {
-            mem: InMemoryVersionedState::new(),
-            height_to_version: HashMap::new(),
-        }
-    }
-
-    pub fn add(&mut self, change: ChangeSet, version: u64) {
-        for (k, v) in Self::convert(change, version) {
+    fn add(&mut self, changes: ChangeSet) {
+        let version = self.next_version();
+        for (k, v) in changes.into_versioned_kv(version) {
             self.mem.historic_state.insert(k, v);
         }
     }
 
-    pub fn set_height_version(&mut self, height: u64, version: u64) {
-        self.height_to_version.insert(height, version);
+    fn tag(&mut self) {
+        self.tags.push(self.version);
     }
 
-    pub fn resolver<'a>(
+    fn resolver<'a>(
         &'a self,
         db: &'a (impl TreeReader<StateKey> + Sync),
-        height: u64,
-    ) -> impl MoveResolver<PartialVMError> + TableResolver + 'a {
-        HistoricResolver::new(
+        height: BlockHeight,
+    ) -> Option<impl MoveResolver<PartialVMError> + TableResolver + 'a> {
+        Some(HistoricResolver::new(
             db,
-            self.height_to_version
-                .get(&height)
-                .copied()
-                .unwrap_or_default(),
+            self.tags.get(height as usize).copied()?,
             &self.mem,
-        )
+        ))
     }
 
-    fn convert(
-        changes: ChangeSet,
-        version: u64,
-    ) -> impl Iterator<Item = ((StateKey, u64), Option<StateValue>)> {
-        changes
-            .into_inner()
+    fn next_version(&mut self) -> Version {
+        self.version += 1;
+        self.version
+    }
+}
+
+/// The `IntoVersionedKv` trait is defined by a single operation [`Self::into_versioned_kv`]. See
+/// its documentation for details.
+trait IntoVersionedKv {
```

Comment:
nit: it should be probably called `IntoVersionedKV` (V is capital) or the long way `IntoVersionedKeyValue`.